### PR TITLE
MSI support, update step detection

### PR DESCRIPTION
**Pull request by hopkiw**
* Add InstallMsi step support
* Eliminate type assertions/type switches in favor of helper functions
* Mark recipes which fail in steps as installed with a fail state
* Add version String() method
* Add debug logging

### CODE REVIEW

File: policies/recipes/steps.go [hunk 1]
```diff
@@ -35,14 +35,14 @@ import (
 	"github.com/ulikunitz/xz/lzma"
 )
 
-// StepFileCopy builds the command for a FileCopy step
-func StepFileCopy(step *osconfigpb.SoftwareRecipe_Step_FileCopy, artifacts map[string]string) error {
-	dest, err := common.NormPath(step.FileCopy.Destination)
+// StepFileCopy runs a FileCopy step
+func StepFileCopy(step *osconfigpb.SoftwareRecipe_Step_CopyFile, artifacts map[string]string, runEnvs []string, stepDir string) error {
```
Comment (adjackura) [inline policies/recipes/steps.go]: StepCopyFile

File: policies/recipes/steps.go [hunk 2]
```diff
@@ -370,51 +372,60 @@ func extractTar(tarName string, dst string, archiveType osconfigpb.SoftwareRecip
 }
 
 // StepMsiInstallation builds the command for a MsiInstallation step
-func StepMsiInstallation(step *osconfigpb.SoftwareRecipe_Step_MsiInstallation, artifacts map[string]string) error {
-	fmt.Println("StepMsiInstallation")
-	return nil
+func StepMsiInstallation(step *osconfigpb.SoftwareRecipe_Step_InstallMsi, artifacts map[string]string, runEnvs []string, stepDir string) error {
```
Comment (adjackura) [inline policies/recipes/steps.go]: Can we name the functions after teh steps, so StepInstallMSI, and StepInstalDPKG

File: policies/recipes/steps.go [hunk 3]
```diff
@@ -370,51 +372,60 @@ func extractTar(tarName string, dst string, archiveType osconfigpb.SoftwareRecip
...
+	artifact := step.GetArtifactId()
+	path, ok := artifacts[artifact]
+	if !ok {
+		return fmt.Errorf("%q not found in artifact map", artifact)
+	}
+	args := step.Flags
+	if len(args) == 0 {
+		args = []string{"/i"}
+	}
+	args = append(args, path)
```
Comment (adjackura) [inline policies/recipes/steps.go]: add these args = /qn, /norestart
and these return codes = 0, 1641, 3010 (even though we say /norestart)

File: policies/recipes/steps.go [hunk 4]
```diff
@@ -370,51 +372,60 @@ func extractTar(tarName string, dst string, archiveType osconfigpb.SoftwareRecip
...
-func StepRpmInstallation(step *osconfigpb.SoftwareRecipe_Step_RpmInstallation, artifacts map[string]string) error {
-	fmt.Println("StepRpmInstallation")
-	return nil
+func StepRpmInstallation(step *osconfigpb.SoftwareRecipe_Step_InstallRpm, artifacts map[string]string, runEnvs []string, stepDir string) error {
+	return fmt.Errorf("InstallRpm not yet supported")
 }
 
 // StepFileExec builds the command for a FileExec step
-func StepFileExec(step *osconfigpb.SoftwareRecipe_Step_FileExec, artifacts map[string]string, runEnvs []string, stepDir string) error {
+func StepFileExec(step *osconfigpb.SoftwareRecipe_Step_ExecFile, artifacts map[string]string, runEnvs []string, stepDir string) error {
```
Comment (adjackura) [inline policies/recipes/steps.go]: StepExecFile

File: policies/recipes/installrecipe.go [hunk 5]
```diff
@@ -37,23 +38,27 @@ func InstallRecipe(ctx context.Context, recipe *osconfigpb.SoftwareRecipe) error
 	}
 	installedRecipe, ok := recipeDB.GetRecipe(recipe.Name)
 	if ok {
-		if (!installedRecipe.Greater(recipe.Version)) &&
-			(recipe.DesiredState == osconfigpb.DesiredState_UPDATED) {
+		logger.Debugf("currently installed version of software recipe %s with version %s", recipe.Name, installedRecipe.version)
```
Comment (adjackura) [inline policies/recipes/installrecipe.go]: Log lines should start with a capitol and end with a period

Comment (hopkiw) [reply policies/recipes/installrecipe.go]: will update

**Pull request merged**